### ISSUE TITLE
Exception thrown from constructor of stateless worker grain eventually freezes grain call

### ISSUE THREAD

**Issue by peter-perot** [dotnet/orleans#5947]
I have found a bug that requires a silo restart to resolve (tested with Orleans versions 1.5.3 and 1.5.9).

You can simply reproduce the bug with the [_HelloWorld_](https://github.com/dotnet/orleans/tree/v1.5.9/Samples/HelloWorld) sample project.

I used the `HelloWorld.SeparateProcess` solution, but I think this doesn't make a difference. At first change the `DoClientWork` method in file `Program.cs` of project `OrleansClient` like this:
```csharp
private static async Task DoClientWork()
{
  var friend = GrainClient.GrainFactory.GetGrain<IHello>(0);

  while (true)
  {
    try
    {
      // example of calling grains from the initialized client
      var response = await friend.SayHello("Good morning, my friend!");
      Console.WriteLine("\n\n{0}\n\n", response);
    }
    catch (Exception exc)
    {
      Console.WriteLine(exc);
    }

    Console.WriteLine("Press RETURN to repeat.");
    Console.ReadLine();
  }
}
```

This will run the grain call in a loop.

Now add a constructor that throws an exception to class `HelloGrain` of project `HelloWorldGrains`:

```csharp
public HelloGrain()
{
  throw new Exception("Intentionally thrown.");
}
```

When running the silo and client, everything works as expected: an `OrleansException`, which contains the thrown exception as an inner exception, is propagated to the client.

So far, so good.

Now make the `HelloGrain` a _stateless worker_:

```csharp
[StatelessWorker]
public class HelloGrain : Orleans.Grain, IHello
{
  // ...
}
```

When running the silo and client, an `OrleansException` is propagated to the client, **but only for 4 iterations**. After that, the client call `var response = await friend.SayHello("Good morning, my friend!");` freezes.

I think the number of 4 has a reason: I run the silo on a Quad Core processor, and as far as I know, Orleans activates a stateless worker with a specific key on each core. Here the key is `0`, which means, 4 grain instances of the stateless worker are tried to instantiate, and then further tries are blocked.

Of course, it is a bad idea to throw an exception from a grain constructor, but we had an issue where this was the case and wondered why our services stopped. Even if the condition that raised the bug in the constructor disappeard, the grain was never again tried to activate.

**FYI**: I tested the same scenario with the `OnActivateAsync` method (throwing an exception from there), but in this case the bug does not occur. So I don't think this bug is related to https://github.com/dotnet/orleans/issues/4748.

**Comment by benjaminpetit** [comment]
Thanks for reporting this, I am able to repro in master too.

**Comment by sergeybykov** [comment]
Fixed via #5958.

**Comment by peter-perot** [comment]
@sergeybykov Thanks! Would be nice if this fix went into 1.5.x, too.

**Comment by sergeybykov** [comment]
Yes, we discussed backporting it to 2.4 and 1.5..
